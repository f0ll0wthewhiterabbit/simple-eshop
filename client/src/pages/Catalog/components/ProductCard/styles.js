import styled from 'styled-components'
import {
  Card,
  CardMedia,
  CardContent,
  CardActions,
  IconButton,
  Button,
  Typography,
} from '@material-ui/core'
import Rating from '@material-ui/lab/Rating'

export const Wrapper = styled(Card)`
  height: 100%;
  display: flex;
  flex-direction: column;
  background: ${props => props.theme.colors.background.productCard};
`
/**
 * 16:9 ratio - 56.25%
 */
export const ImageWrapper = styled(CardMedia)`
  padding-top: 56.25%;
`

export const Content = styled(CardContent)`
  flex-grow: 1;
  position: relative;
  padding-bottom: 18px;
`

export const Title = styled(Typography)`
  padding-right: 70px;
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.bold};
  font-weight: 700;
  text-transform: none;
  font-size: 20px;
`

export const Price = styled(Typography)`
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.bold};
  font-weight: 700;
  text-transform: none;
  font-size: 16px;
  margin-bottom: 25px;
`

export const Description = styled(Typography)`
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.regular};
  font-weight: 500;
  text-transform: none;
  font-size: 16px;
`

export const PurchaseButton = styled(Button)`
  font-family: ${props => props.theme.font.family};
  font-size: 13px;
  font-weight: 600;
  position: absolute;
  top: 16px;
  right: 16px;
  margin-right: 4px;
`

export const TagsWrapper = styled(CardActions)`
  padding-left: ${props => props.theme.spacing(2)}px;
  padding-right: ${props => props.theme.spacing(2)}px;
  flex-wrap: wrap;

  .MuiChip-root {
    margin-left: 0;
    margin-right: 6px;
    margin-bottom: 2px;
    padding: 5px 10px;
    font-family: ${props => props.theme.font.family};
    font-size: 12px;
    background: ${props => props.theme.colors.accent};
    color: ${props => props.theme.colors.font.contrast};
    border-radius: 40px;
  }
`

export const RaitingsRoot = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: ${props => props.theme.spacing(2, 2)};
`

export const RaitingWrapper = styled(CardActions)`
  display: flex;
  justify-content: flex-start;
  flex-wrap: wrap;
  padding: 0;
  width: 50%;
  min-height: 42px;
  position: relative;
`

export const Stars = styled(Rating)`
  min-height: 18px;

  .MuiRating-iconEmpty {
    color: ${props => props.theme.colors.font.light};
  }

  && {
    margin-left: 0;
  }
`

export const RaitingsCount = styled.span`
  margin-left: 6px;
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.light};
  font-weight: 500;
  font-size: 12px;
`

export const DeleteButton = styled(IconButton)`
  color: ${props => props.theme.colors.font.light};
  position: absolute;
  top: 19px;
  left: 87px;

  && {
    margin-left: 3px;
  }

  &:hover {
    color: ${props => props.theme.colors.font.regular};
  }
`

export const RatingTitle = styled.span`
  width: 100%;
  margin-left: 2px;
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.light};
  font-weight: 500;
  font-size: 12px;
`

export const LoadingLabel = styled(Typography)`
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.light};
  font-weight: 500;
  font-size: 12px;
  margin-left: 2px;
`

export const ErrorLabel = styled(Typography)`
  margin-left: 2px;
  font-family: ${props => props.theme.font.family};
  font-weight: 500;
  font-size: 12px;
`
