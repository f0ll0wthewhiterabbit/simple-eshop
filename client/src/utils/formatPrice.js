const formatPrice = price => {
  const USDFormatter = new Intl.NumberFormat('ru', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: price < 1000 ? 2 : 0,
  })

  return USDFormatter.format(price)
}

export default formatPrice
