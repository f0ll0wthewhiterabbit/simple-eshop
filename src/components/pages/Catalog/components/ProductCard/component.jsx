import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Typography, Button, Chip } from '@material-ui/core'
import ShoppingCartOutlinedIcon from '@material-ui/icons/ShoppingCartOutlined'
import CancelIcon from '@material-ui/icons/Cancel'

import {
  Wrapper,
  ImageWrapper,
  Content,
  TagsWrapper,
  ActionsWrapper,
  Stars,
  RaitingWrapper,
  RaitingsCount,
  DeleteButton,
  RatingCount,
} from './styles'

const ProductCard = ({ productData, currentUserId, changeProductRating, deleteProductRating }) => {
  const [isUserChangedRatign, setIsUserChangedRating] = useState(false)

  const { id, title, description, tags, price, rating, image: imageSrc } = productData
  const ratingsAmount = rating.length
  const averageRating = Math.round(rating.reduce((a, b) => a + b.stars, 0) / ratingsAmount)
  const currentUserRating = rating.find(it => it.userId === currentUserId)
  const isUserRatedProduct = Boolean(currentUserRating)

  const handleRatingChange = (event, userRating) => {
    changeProductRating({
      productId: id,
      userRating,
    })

    setTimeout(() => {
      setIsUserChangedRating(true)
    }, 500)
  }

  const handleDeleteButtonClick = () => {
    deleteProductRating(id)
    setIsUserChangedRating(false)
  }

  return (
    <Wrapper>
      <ImageWrapper image={imageSrc} title={title} />
      <Content>
        <Typography gutterBottom variant="h5" component="h2">
          {title}
        </Typography>
        <Typography>{description}</Typography>
      </Content>
      <TagsWrapper>
        {tags.map(tag => (
          <Chip key={tag} label={tag} color="secondary" size="small" />
        ))}
      </TagsWrapper>
      <ActionsWrapper>
        <RaitingWrapper>
          {isUserRatedProduct && <RatingCount>Your rating: {currentUserRating.stars}</RatingCount>}
          <Stars
            value={averageRating}
            name={`simple-controlled-${id}`}
            size="small"
            onChange={handleRatingChange}
            readOnly={isUserChangedRatign}
          />{' '}
          <RaitingsCount data-color={isUserRatedProduct ? '#f50057' : '#bdbdbd'} active>
            ({ratingsAmount})
          </RaitingsCount>
          {isUserRatedProduct && (
            <DeleteButton aria-label="delete" size="small" onClick={handleDeleteButtonClick}>
              <CancelIcon fontSize="inherit" />
            </DeleteButton>
          )}
        </RaitingWrapper>
        <Button size="small" color="primary" startIcon={<ShoppingCartOutlinedIcon />} disabled>
          {price} $
        </Button>
      </ActionsWrapper>
    </Wrapper>
  )
}

ProductCard.propTypes = {
  productData: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    tags: PropTypes.arrayOf(PropTypes.string).isRequired,
    price: PropTypes.number.isRequired,
    rating: PropTypes.arrayOf(
      PropTypes.shape({
        userId: PropTypes.number.isRequired,
        stars: PropTypes.number.isRequired,
      })
    ).isRequired,
    image: PropTypes.string.isRequired,
  }).isRequired,
  currentUserId: PropTypes.number.isRequired,
  changeProductRating: PropTypes.func.isRequired,
  deleteProductRating: PropTypes.func.isRequired,
}

export default ProductCard
