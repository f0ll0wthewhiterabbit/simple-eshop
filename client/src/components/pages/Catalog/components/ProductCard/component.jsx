import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { Typography, Chip } from '@material-ui/core'
import ShoppingCartOutlinedIcon from '@material-ui/icons/ShoppingCartOutlined'
import DeleteIcon from '@material-ui/icons/Delete'

import formatPrice from '../../../../../utils/formatPrice'
import { API_URL } from '../../../../../constants'
import {
  Wrapper,
  ImageWrapper,
  Content,
  Title,
  PriceButton,
  TagsWrapper,
  Stars,
  RaitingsRoot,
  RaitingWrapper,
  RaitingsCount,
  DeleteButton,
  RatingTitle,
  LoadingLabel,
  ErrorLabel,
} from './styles'

const ProductCard = ({
  productData,
  currentUserId,
  changeProductRating,
  deleteProductRating,
  ratingsLoadingList,
  ratingsErrorList,
}) => {
  const { id, title, description, tags, price, rating, imageName } = productData
  const ratingsAmount = rating.size
  const averageRating = Math.round(rating.reduce((a, b) => a + b.stars, 0) / ratingsAmount)
  const currentUserRating = rating.find(it => it.userId === currentUserId)
  const isUserRatedProduct = Boolean(currentUserRating)
  const isRatingLoading = ratingsLoadingList.indexOf(id) !== -1
  const isErrorInLoad = ratingsErrorList.indexOf(id) !== -1

  const handleRatingChange = (event, userRating) => {
    changeProductRating({
      productId: id,
      userRating,
    })
  }

  const handleDeleteButtonClick = () => {
    deleteProductRating(id)
  }

  let userRatingField

  if (isErrorInLoad) {
    userRatingField = (
      <ErrorLabel variant="caption" color="secondary">
        Something went wrong!
      </ErrorLabel>
    )
  } else if (isRatingLoading) {
    userRatingField = <LoadingLabel variant="caption">Loading...</LoadingLabel>
  } else {
    userRatingField = (
      <>
        <Stars
          value={isUserRatedProduct ? currentUserRating.stars : 0}
          name={`simple-controlled-user-${id}`}
          size="small"
          onChange={handleRatingChange}
          readOnly={isUserRatedProduct}
        />
        {isUserRatedProduct && (
          <DeleteButton aria-label="delete" size="small" onClick={handleDeleteButtonClick}>
            <DeleteIcon fontSize="inherit" />
          </DeleteButton>
        )}
      </>
    )
  }

  return (
    <Wrapper>
      <ImageWrapper image={`${API_URL}/products/${id}/${imageName}`} title={title} />
      <Content>
        <Title gutterBottom variant="h5" component="h2">
          {title}
        </Title>
        <PriceButton size="small" color="primary" startIcon={<ShoppingCartOutlinedIcon />} disabled>
          {formatPrice(price)}
        </PriceButton>
        <Typography>{description}</Typography>
      </Content>
      <TagsWrapper>
        {tags.map(tag => (
          <Chip key={tag} label={tag} color="secondary" size="small" />
        ))}
      </TagsWrapper>
      <RaitingsRoot>
        <RaitingWrapper>
          <RatingTitle>Your rating:</RatingTitle>
          {userRatingField}
        </RaitingWrapper>
        <RaitingWrapper>
          <RatingTitle>Average rating:</RatingTitle>
          <Stars
            value={averageRating}
            name={`simple-controlled-average-${id}`}
            onChange={handleRatingChange}
            readOnly
            size="small"
          />
          <RaitingsCount active>({ratingsAmount})</RaitingsCount>
        </RaitingWrapper>
      </RaitingsRoot>
    </Wrapper>
  )
}

ProductCard.defaultProps = {
  currentUserId: '',
}

ProductCard.propTypes = {
  productData: ImmutablePropTypes.recordOf({
    _id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    imageName: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    tags: ImmutablePropTypes.listOf(PropTypes.string),
    rating: ImmutablePropTypes.listOf(
      ImmutablePropTypes.recordOf({
        _id: PropTypes.string.isRequired,
        userId: PropTypes.string.isRequired,
        stars: PropTypes.number.isRequired,
      })
    ),
  }).isRequired,
  currentUserId: PropTypes.string,
  changeProductRating: PropTypes.func.isRequired,
  deleteProductRating: PropTypes.func.isRequired,
  ratingsLoadingList: ImmutablePropTypes.listOf(PropTypes.string).isRequired,
  ratingsErrorList: ImmutablePropTypes.listOf(PropTypes.string).isRequired,
}

export default ProductCard
