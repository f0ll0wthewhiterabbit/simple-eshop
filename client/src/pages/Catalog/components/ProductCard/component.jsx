import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { Chip } from '@material-ui/core'
import ShoppingCartOutlinedIcon from '@material-ui/icons/ShoppingCartOutlined'

import Ratings from './components/Ratings'
import formatPrice from '../../../../utils/formatPrice'
import { BASE_URL } from '../../../../utils/api'
import {
  Wrapper,
  ImageWrapper,
  Content,
  Title,
  PurchaseButton,
  TagsWrapper,
  Price,
  Description,
} from './styles'

const ProductCard = ({ productData }) => {
  const { _id: id, title, description, tags, price, imageName } = productData

  return (
    <Wrapper>
      <ImageWrapper image={`${BASE_URL}/products/${id}/${imageName}`} title={title} />
      <Content>
        <Title gutterBottom variant="h5" component="h2" data-test="title">
          {title}
        </Title>
        <Price gutterBottom variant="h5" component="h3">
          {formatPrice(price)}
        </Price>
        <PurchaseButton size="small" color="secondary" startIcon={<ShoppingCartOutlinedIcon />}>
          Buy
        </PurchaseButton>
        <Description data-test="description">{description}</Description>
      </Content>
      <TagsWrapper>
        {tags.map(tag => (
          <Chip key={tag} label={tag} color="secondary" size="small" data-test="tag" />
        ))}
      </TagsWrapper>
      <Ratings productData={productData} />
    </Wrapper>
  )
}

ProductCard.propTypes = {
  productData: ImmutablePropTypes.recordOf({
    _id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    imageName: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    tags: ImmutablePropTypes.listOf(PropTypes.string),
    ratingInfo: ImmutablePropTypes.recordOf({
      average: PropTypes.oneOfType([
        PropTypes.number.isRequired,
        PropTypes.oneOf([null]).isRequired,
      ]),
      votesAmount: PropTypes.number.isRequired,
      currentUserRating: PropTypes.number,
    }),
  }).isRequired,
}

export default ProductCard
