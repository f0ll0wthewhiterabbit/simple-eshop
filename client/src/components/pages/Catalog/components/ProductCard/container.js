import { connect } from 'react-redux'

import ProductCard from './component'
import { changeProductRating, deleteProductRating } from '../../../../../store/actions'

const mapStateToProps = state => ({
  currentUserId: state.users.current.id,
})

const mapDispatchToProps = dispatch => ({
  changeProductRating: productData => dispatch(changeProductRating(productData)),
  deleteProductRating: productId => dispatch(deleteProductRating(productId)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard)
