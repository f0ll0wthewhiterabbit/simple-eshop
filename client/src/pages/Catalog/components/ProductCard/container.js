import { connect } from 'react-redux'

import ProductCard from './component'
import { changeProductRatingRequest, deleteProductRatingRequest } from '../../../../store/actions'

const mapStateToProps = state => ({
  ratingsLoadingList: state.getIn(['products', 'ratingsLoadingList']),
  ratingsErrorList: state.getIn(['products', 'ratingsErrorList']),
})

const mapDispatchToProps = {
  changeProductRatingRequest: productData => changeProductRatingRequest(productData),
  deleteProductRatingRequest: productId => deleteProductRatingRequest(productId),
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard)
