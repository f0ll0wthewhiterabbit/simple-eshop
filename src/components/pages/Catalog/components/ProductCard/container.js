import { connect } from 'react-redux'

import ProductCard from './component'
import { changeProductRating } from '../../../../../store/actions/products'

const mapStateToProps = state => ({
  currentUserId: state.users.current.id,
})

const mapDispatchToProps = dispatch => ({
  changeProductRating: productData => dispatch(changeProductRating(productData)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductCard)
