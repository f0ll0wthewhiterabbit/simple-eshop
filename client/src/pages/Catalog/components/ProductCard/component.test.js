import React from 'react'
import { shallow } from 'enzyme'
import { Record, List } from 'immutable'

import ProductCard from './component'
import findByTestAttr from '../../../../utils/findByTestAttr'
import formatPrice from '../../../../utils/formatPrice'

describe('ProductCard component', () => {
  let wrapper
  const productData = Record({
    _id: '1',
    title: 'TestTitle',
    description: 'TestDescription',
    imageName: 'TestImageName',
    price: 99,
    tags: List(['firstTag', 'secondTag']),
    ratingInfo: Record({
      average: 5,
      votesAmount: 2,
      currentUserRating: 5,
    })(),
  })()
  const initialProps = {
    productData,
    changeProductRating: jest.fn(),
    deleteProductRating: jest.fn(),
    ratingsLoadingList: List(),
    ratingsErrorList: List(),
  }
  const dataTestTitle = 'title'
  const dataTestDescription = 'description'
  const dataTestTag = 'tag'
  const dataTestPrice = 'price'
  const dataTestAverageRating = 'averageRating'
  const dataTestRatingsCount = 'ratingsCount'
  const dataTestUserRating = 'userRating'
  const dataTestRatingError = 'ratingError'
  const dataTestRatingLoading = 'ratingLoading'
  const dataTestDeleteButton = 'deleteButton'

  const generateWrapper = passedProps => {
    const defaultProps = { ...initialProps }
    const props = { ...defaultProps, ...passedProps }

    return shallow(<ProductCard {...props} />)
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  afterEach(() => {
    initialProps.changeProductRating.mockReset()
    initialProps.deleteProductRating.mockReset()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should contain correct title', () => {
    const title = findByTestAttr(wrapper, dataTestTitle)

    expect(title.text()).toBe(productData.title)
  })

  it('should contain correct description', () => {
    const description = findByTestAttr(wrapper, dataTestDescription)

    expect(description.text()).toBe(productData.description)
  })

  it('should contain correct price', () => {
    const price = findByTestAttr(wrapper, dataTestPrice)
    const expectedPrice = formatPrice(productData.price)

    expect(price.text()).toBe(expectedPrice)
  })

  it('should contain correct tags', () => {
    const tags = findByTestAttr(wrapper, dataTestTag)

    expect(tags).toHaveLength(productData.tags.size)
  })

  it('should contain correct average rating', () => {
    const averageRating = findByTestAttr(wrapper, dataTestAverageRating)

    expect(averageRating.props().value).toBe(productData.ratingInfo.average)
  })

  it('should contain correct votes amount', () => {
    const ratingsCount = findByTestAttr(wrapper, dataTestRatingsCount)
    const expectedAmount = `(${productData.ratingInfo.votesAmount})`

    expect(ratingsCount.text()).toBe(expectedAmount)
  })

  it('should contain correct user rating', () => {
    const userRating = findByTestAttr(wrapper, dataTestUserRating)

    expect(userRating.props().value).toBe(productData.ratingInfo.currentUserRating)
  })

  it('should contain delete user rating button', () => {
    const deleteButton = findByTestAttr(wrapper, dataTestDeleteButton)

    expect(deleteButton).toHaveLength(1)
  })

  it('should handle delete user rating', () => {
    const deleteButton = findByTestAttr(wrapper, dataTestDeleteButton)
    deleteButton.simulate('click')

    expect(initialProps.deleteProductRating).toHaveBeenCalledTimes(1)
    expect(initialProps.deleteProductRating).toHaveBeenCalledWith(productData._id)
  })

  it('should handle change user rating', () => {
    const newUserRating = 1
    const userRating = findByTestAttr(wrapper, dataTestUserRating)
    userRating.simulate('change', {}, newUserRating)

    expect(initialProps.changeProductRating).toHaveBeenCalledTimes(1)
    expect(initialProps.changeProductRating).toHaveBeenCalledWith({
      productId: productData._id,
      userRating: newUserRating,
    })
  })

  it('should contain correct raiting error message', () => {
    wrapper = generateWrapper({ ratingsErrorList: List([productData._id]) })
    const ratingError = findByTestAttr(wrapper, dataTestRatingError)

    expect(ratingError).toHaveLength(1)
    expect(ratingError.text()).toBe('Something went wrong!')
  })

  it('should contain correct raiting loading message', () => {
    wrapper = generateWrapper({ ratingsLoadingList: List([productData._id]) })
    const ratingLoading = findByTestAttr(wrapper, dataTestRatingLoading)

    expect(ratingLoading).toHaveLength(1)
    expect(ratingLoading.text()).toBe('Loading...')
  })

  describe('without user rating', () => {
    beforeEach(() => {
      const productDataWithoutUserRating = Record({
        _id: '1',
        title: 'TestTitle',
        description: 'TestDescription',
        imageName: 'TestImageName',
        price: 99,
        tags: List(['firstTag', 'secondTag']),
        ratingInfo: Record({
          average: 5,
          votesAmount: 1,
        })(),
      })()
      wrapper = generateWrapper({ productData: productDataWithoutUserRating })
    })

    it(`shouldn't contain user rating`, () => {
      const userRating = findByTestAttr(wrapper, dataTestUserRating)

      expect(userRating.props().value).toBe(0)
    })

    it(`shouldn't contain delete user rating button`, () => {
      const deleteButton = findByTestAttr(wrapper, dataTestDeleteButton)

      expect(deleteButton).toHaveLength(0)
    })
  })
})
