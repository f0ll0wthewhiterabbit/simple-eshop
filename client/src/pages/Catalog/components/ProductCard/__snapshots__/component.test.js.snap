// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component should render correctly 1`] = `
<Styled(WithStyles(ForwardRef(Card)))>
  <Styled(WithStyles(ForwardRef(CardMedia)))
    image="http://localhost:5000/api/products/1/TestImageName"
    title="TestTitle"
  />
  <Styled(WithStyles(ForwardRef(CardContent)))>
    <Styled(WithStyles(ForwardRef(Typography)))
      component="h2"
      data-test="title"
      gutterBottom={true}
      variant="h5"
    >
      TestTitle
    </Styled(WithStyles(ForwardRef(Typography)))>
    <Styled(WithStyles(ForwardRef(Typography)))
      component="h3"
      gutterBottom={true}
      variant="h5"
    />
    <Styled(WithStyles(ForwardRef(Button)))
      color="secondary"
      size="small"
      startIcon={<Memo(ShoppingCartOutlinedIcon) />}
    >
      Buy
    </Styled(WithStyles(ForwardRef(Button)))>
    <Styled(WithStyles(ForwardRef(Typography)))
      data-test="description"
    >
      TestDescription
    </Styled(WithStyles(ForwardRef(Typography)))>
  </Styled(WithStyles(ForwardRef(CardContent)))>
  <Styled(WithStyles(ForwardRef(CardActions)))>
    <WithStyles(ForwardRef(Chip))
      color="secondary"
      data-test="tag"
      key="firstTag"
      label="firstTag"
      size="small"
    />
    <WithStyles(ForwardRef(Chip))
      color="secondary"
      data-test="tag"
      key="secondTag"
      label="secondTag"
      size="small"
    />
  </Styled(WithStyles(ForwardRef(CardActions)))>
  <styled.div>
    <Styled(WithStyles(ForwardRef(CardActions)))>
      <styled.span>
        Your rating:
      </styled.span>
      <Styled(WithStyles(ForwardRef(Rating)))
        data-test="userRating"
        name="simple-controlled-user-1"
        onChange={[Function]}
        readOnly={true}
        size="small"
        value={5}
      />
      <Styled(WithStyles(ForwardRef(IconButton)))
        aria-label="delete"
        data-test="deleteButton"
        onClick={[Function]}
        size="small"
      >
        <DeleteIcon
          fontSize="inherit"
        />
      </Styled(WithStyles(ForwardRef(IconButton)))>
    </Styled(WithStyles(ForwardRef(CardActions)))>
    <Styled(WithStyles(ForwardRef(CardActions)))>
      <styled.span>
        Average rating:
      </styled.span>
      <Styled(WithStyles(ForwardRef(Rating)))
        data-test="averageRating"
        name="simple-controlled-average-1"
        onChange={[Function]}
        readOnly={true}
        size="small"
        value={5}
      />
      <styled.span
        active={true}
        data-test="ratingsCount"
      >
        (
        2
        )
      </styled.span>
    </Styled(WithStyles(ForwardRef(CardActions)))>
  </styled.div>
</Styled(WithStyles(ForwardRef(Card)))>
`;
